src/proxy/ssh/server.ts

```diff
@@ -0,0 +1,623 @@
+    if (!Number.isFinite(bytes) || bytes <= 0) {
+      return `${bytes} bytes`;
+    }
+
+    const units = ['bytes', 'KB', 'MB', 'GB', 'TB'];
+    let value = bytes;
+    let unitIndex = 0;
+
+    while (value >= 1024 && unitIndex < units.length - 1) {
+      value /= 1024;
+      unitIndex++;
+    }
+
+    const precision = unitIndex === 0 ? 0 : 2;
+    return `${value.toFixed(precision)} ${units[unitIndex]}`;
+  }
+
+  async handleClient(
+    client: ssh2.Connection,
+    clientInfo?: { ip?: string; family?: string },
+  ): Promise<void> {
+    const clientIp = clientInfo?.ip || 'unknown';
+    console.log(`[SSH] Client connected from ${clientIp}`);
+    const clientWithUser = client as ClientWithUser;
+    clientWithUser.clientIp = clientIp;
+
+    const connectionTimeout = setTimeout(() => {
+      console.log(`[SSH] Connection timeout for ${clientIp} - closing`);
+      client.end();
+    }, 600000); // 10 minute timeout
+
+    client.on('error', (err: Error) => {
+      console.error(`[SSH] Client error from ${clientIp}:`, err);
+      clearTimeout(connectionTimeout);
+    });
+
+    client.on('end', () => {
+      console.log(`[SSH] Client disconnected from ${clientIp}`);
+      clearTimeout(connectionTimeout);
+    });
+
+    client.on('close', () => {
+      console.log(`[SSH] Client connection closed from ${clientIp}`);
+      clearTimeout(connectionTimeout);
+    });
+
+    (client as any).on('global request', (accept: () => void, reject: () => void, info: any) => {
+      if (info.type === 'keepalive@openssh.com') {
+        accept();
+      } else {
+        reject();
+      }
+    });
+
+    client.on('authentication', (ctx: ssh2.AuthContext) => {
+      console.log(
+        `[SSH] Authentication attempt from ${clientIp}:`,
+        ctx.method,
+        'for user:',
+        ctx.username,
+      );
+
+      if (ctx.method === 'publickey') {
+        const keyString = `${ctx.key.algo} ${ctx.key.data.toString('base64')}`;
+        console.log(
+          '[SSH] Attempting to find user by SSH key: ',
+          JSON.stringify(keyString, null, 2),
+        );
+
+        db.findUserBySSHKey(keyString)
+          .then((user: any) => {
+            if (user) {
+              console.log(
+                `[SSH] Public key authentication successful for user: ${user.username} from ${clientIp}`,
+              );
+              clientWithUser.authenticatedUser = {
+                username: user.username,
+                email: user.email,
+                gitAccount: user.gitAccount,
+              };
+              ctx.accept();
+            } else {
+              console.log('[SSH] Public key authentication failed - key not found');
+              ctx.reject();
+            }
+          })
+          .catch((err: Error) => {
+            console.error('[SSH] Database error during public key auth:', err);
+            ctx.reject();
+          });
+      } else if (ctx.method === 'password') {
```

Comment:
I have a feeling this should be removed. Apparently this is a fallback for when there are no valid keys present in the user's machine. (?) The user sees a `git@<host-ip>'s password:` prompt that asks for the password, and the `ctx.username` value here would be `git`. Of course, this fails if `git` is not a GitProxy user. This is why my flow was passing the check when using `admin@localhost` with the `admin` password.

I haven't been able to send a message to the user that only public key auth is possible. Wonder if this is possible at all? Simply throwing errors or emitting error events doesn't seem to work.